### DISCUSSION TITLE
App kept reporting timeouts (stack gunicorn + uvicorn + nginx)

### DISCUSSION THREAD

**Discussion by thaodt** [fastapi/fastapi#6684] [Questions]
### First check
* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description
Hi guys, seriously I need your helps. Currently my app is running under nginx load balancer + gunicorn & uvicorn workers.
This is my current configuration:
```2021-06-19 06:59:16.478 | DEBUG    | gunicorn.glogging:debug:265 - Current configuration:
  config: gunicorn_conf.py
  bind: ['0.0.0.0:8081']
  backlog: 2048
  workers: 8
  worker_class: uvicorn.workers.UvicornWorker
  threads: 1
  worker_connections: 1000
  max_requests: 0
  max_requests_jitter: 0
  timeout: 360
  graceful_timeout: 60
  keepalive: 120
  limit_request_line: 4094
  limit_request_fields: 100
  limit_request_field_size: 8190
  reload: False
  reload_engine: auto
  reload_extra_files: []
  spew: False
  check_config: False
  preload_app: True
  sendfile: None
  reuse_port: False
  chdir: /app
  daemon: False
  raw_env: []
  pidfile: None
  worker_tmp_dir: None
  user: 0
  group: 0
  umask: 0
  initgroups: False
  tmp_upload_dir: None
  secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
  forwarded_allow_ips: ['127.0.0.1']
  accesslog: None
  disable_redirect_access_to_syslog: False
  access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
  errorlog: -
  loglevel: DEBUG
  capture_output: False
  logger_class: gunicorn.glogging.Logger
  logconfig: None
  logconfig_dict: {}
  syslog_addr: udp://localhost:514
  syslog: False
  syslog_prefix: None
  syslog_facility: user
  enable_stdio_inheritance: False
  statsd_host: None
  dogstatsd_tags: 
  statsd_prefix: 
  proc_name: None
  default_proc_name: mb_driver.main:app
  pythonpath: None
  paste: None
  on_starting: <function OnStarting.on_starting at 0x7f9964dc0790>
  on_reload: <function OnReload.on_reload at 0x7f9964dc08b0>
  when_ready: <function WhenReady.when_ready at 0x7f9964dc09d0>
  pre_fork: <function Prefork.pre_fork at 0x7f9964dc0af0>
  post_fork: <function Postfork.post_fork at 0x7f9964dc0c10>
  post_worker_init: <function PostWorkerInit.post_worker_init at 0x7f9964dc0d30>
  worker_int: <function WorkerInt.worker_int at 0x7f9964dc0e50>
  worker_abort: <function WorkerAbort.worker_abort at 0x7f9964dc0f70>
  pre_exec: <function PreExec.pre_exec at 0x7f99647880d0>
  pre_request: <function PreRequest.pre_request at 0x7f99647881f0>
  post_request: <function PostRequest.post_request at 0x7f9964788280>
  child_exit: <function ChildExit.child_exit at 0x7f99647883a0>
  worker_exit: <function WorkerExit.worker_exit at 0x7f99647884c0>
  nworkers_changed: <function NumWorkersChanged.nworkers_changed at 0x7f99647885e0>
  on_exit: <function OnExit.on_exit at 0x7f9964788700>
  proxy_protocol: False
  proxy_allow_ips: ['127.0.0.1']
  keyfile: None
  certfile: None
  ssl_version: 2
  cert_reqs: 0
  ca_certs: None
  suppress_ragged_eofs: True
  do_handshake_on_connect: False
  ciphers: None
  raw_paste_global_conf: []
  strip_header_spaces: False
```
On nginx conf, i configured proxy timeouts as some suggestions on google & our github issuezs here:
```
proxy_redirect off;
proxy_set_header Host $host;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_intercept_errors on;
proxy_connect_timeout 75s;
proxy_read_timeout 300s;
proxy_send_timeout 300s;
```
timeout on gunicorn was set `360`, timeout on nginx was `300`. 

And just sometimes I got WORKER TIMEOUT. 
When I checked my nginx error logs, it threw a ton of error like below:
```
2021/06/19 15:19:17 [error] 13073#13073: *11687 upstream prematurely closed connection while reading response header from upstream, client: xxxxxxxxxx, server: xxxxxxxxxx, request: "GET /xxxxxxxxxx HTTP/1.1", upstream: "http://127.0.0.1:8081/xxxxxxxxxx", host: "xxxxxxxxxx"
2021/06/19 15:22:37 [info] 13074#13074: *12061 epoll_wait() reported that client prematurely closed connection, so upstream connection is closed too while sending request to upstream, client: xxxxxxxxxx, server: xxxxxxxxxx, request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8081/", host: "xxxxxxxxxx"
2021/06/19 15:24:21 [error] 13073#13073: *11823 upstream timed out (110: Connection timed out) while reading response header from upstream, client: xxxxxxxxxx server: xxxxxxxxxx, request: "GET /xxxxxxxxx HTTP/1.1", upstream: "http://127.0.0.1:8081/xxxxxxx", host: "xxxxxxxxxx"
```

### Environment

* OS: [e.g. Linux / Windows / macOS]: CentOS Linux 7 (Core) 
* Kernel: Linux 3.10.0-1160.el7.x86_64
* My required packages:
```
fastapi==0.61.1
gunicorn==20.0.4
uvicorn[standard]==0.12.3
rethinkdb==2.4.7
```
I don't know what to do now. And my nginx access logs are a lot of 502 / 504 / 499 response codes.

Please help!

### Additional context
My rethinkdb server setup was 8 cpu and 32 memory. Sometimes app was made it high CPU load.

**Comment by Mause** [comment]
Are you sure it isn't your app being really slow?

**Comment by thaodt** [comment]
> Are you sure it isn't your app being really slow?

Hi @Mause, so i suppose all my configurations above are okay. 
For your question, I dont think so. My apis called to 3rd party api, get its response and log into rethinkdb. 
Logging into db is a background task feature of FastAPI.
So data will be returned back to client without waiting db insertion complete.

But I noticed some responses with 502 or 499 when calling to 3rd party apis. It could be up to more than 90s to get response from 3rd apis.
Do you think this one led to my upstream got timeouts? The problem is caused by 3rd party apis?

**Comment by thaodt** [comment]
Close this ticket. Issue was resolved. Thank for all helps.

**Comment by monyuan** [comment]
> 关闭这张票。问题已解决。感谢所有帮助。

请问是如何解决的
